cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -493,9 +493,17 @@ public void programHistory(HttpRequest request, HttpResponder responder,
                              @QueryParam("end") String endTs,
                              @QueryParam("limit") @DefaultValue("100") final int resultLimit)
     throws Exception {
-    programHistoryVersioned(request, responder, namespaceId, appName,
-                   getLatestAppVersion(new NamespaceId(namespaceId), appName), type,
-                   programName, status, startTs, endTs, resultLimit);
+    ProgramType programType = getProgramType(type);
+
+    long start = (startTs == null || startTs.isEmpty()) ? 0 : Long.parseLong(startTs);
+    long end = (endTs == null || endTs.isEmpty()) ? Long.MAX_VALUE : Long.parseLong(endTs);
+    ProgramRunStatus runStatus = (status == null) ? ProgramRunStatus.ALL :
+      ProgramRunStatus.valueOf(status.toUpperCase());
+
+    ProgramReference programReference = new ProgramReference(namespaceId, appName, programType, programName);
+    List<RunRecord> records = lifecycleService.getAllRunRecords(programReference, runStatus, start, end, resultLimit);
+    records = records.stream().filter(record -> !isTetheredRunRecord(record)).collect(Collectors.toList());
```

Comment:
this looks like it may have been an existing problem, but it's not good to filter and collect in another pass. This requires an extra pass through the data and also doubles the memory usage. This is especially important when the method can return a large collection like this one.

Instead, we should have a version of the getAllRunRecords() method that also takes a `Predicate`, so that it can filter during the first pass.

Reply:
added a predicate filter